sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx

```diff
@@ -0,0 +1,192 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=True
+
```

Comment:
Please document the classes in this file in detail, including concurrency / locking expectations, and a sketch for the algorithm if possible.
I'll continue to review these in the meantime.

Reply:
I'll add them in next revision. For sketch of the algorithm, maybe it's better to demo in unit tests. I have a typical usage test case in corresponding unit test and I'll add more comments in that unit test.